src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToComponentCodeActionResolver.cs

```diff
@@ -146,4 +182,759 @@ internal sealed class ExtractToComponentCodeActionResolver
             DocumentChanges = documentChanges,
         };
     }
+
+    private static ExtractToComponentCodeActionParams? DeserializeActionParams(JsonElement data)
+    {
+        return data.ValueKind == JsonValueKind.Undefined
+            ? null
+            : JsonSerializer.Deserialize<ExtractToComponentCodeActionParams>(data.GetRawText());
+    }
+
+    internal sealed record SelectionAnalysisResult
+    {
+        public required bool Success;
+        public int ExtractStart;
+        public int ExtractEnd;
+        public bool HasAtCodeBlock;
+        public bool HasEventHandlerOrExpression;
+        public HashSet<string>? UsingDirectives;
+    }
+
+    private static SelectionAnalysisResult TryAnalyzeSelection(RazorCodeDocument codeDocument, ExtractToComponentCodeActionParams actionParams)
+    {
+        var syntaxTree = codeDocument.GetSyntaxTree();
+        var sourceText = codeDocument.Source.Text;
+
+        var (startElementNode, endElementNode) = GetStartAndEndElements(sourceText, syntaxTree, actionParams);
+        if (startElementNode is null)
+        {
+            return new SelectionAnalysisResult { Success = false };
+        }
+
+        endElementNode ??= startElementNode;
+        
+        var success = TryProcessSelection(startElementNode,
+            endElementNode,
+            codeDocument,
+            actionParams,
+            out var extractStart,
+            out var extractEnd,
+            out var hasAtCodeBlock);
+
+        if (!success)
+        {
+            return new SelectionAnalysisResult { Success = false };
+        }
+
+        var dependencyScanRoot = FindNearestCommonAncestor(startElementNode, endElementNode) ?? startElementNode;
+        var usingDirectives = GetUsingDirectivesInRange(syntaxTree, dependencyScanRoot, extractStart, extractEnd);
+        var hasOtherIdentifiers = CheckHasOtherIdentifiers(dependencyScanRoot, extractStart, extractEnd);
+
+        return new SelectionAnalysisResult
+        {
+            Success = success,
+            ExtractStart = extractStart,
+            ExtractEnd = extractEnd,
+            HasAtCodeBlock = hasAtCodeBlock,
+            HasEventHandlerOrExpression = hasOtherIdentifiers,
+            UsingDirectives = usingDirectives,
+        };
+    }
+
+    private static (MarkupSyntaxNode? Start, MarkupSyntaxNode? End) GetStartAndEndElements(SourceText sourceText, RazorSyntaxTree syntaxTree, ExtractToComponentCodeActionParams actionParams)
+    {
+        if (syntaxTree is null)
+        {
+            return (null, null);
+        }
+
+        var owner = syntaxTree.Root.FindInnermostNode(actionParams.AbsoluteIndex, includeWhitespace: true);
```

Comment:
We're doing this twice. Why can't we provide the correct way to get the start and end in the provider? We already have to check the nodes there anyways

Reply:
Would it be an improvement if I pass in the spans of the start and element nodes to the resolver and use `FindNode()` on those spans? 

If not, I think the way to go would be to move all of the `TryAnalyzeSelection()` stuff to the provider.

Reply:
I would say any analysis of validation of selection should always go in the provider, and the action of determining work beyond that is in the resolver. Keep in mind failure in a resolver means that an action was shown and just fails to apply for some reason and isn't as obvious to a user.

Reply:
Yep. Working on moving selection validation to Provider again